[RFC] Fix arrayJoin() capturing in lambda

azat opened the pull request:
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix arrayJoin() capturing in lambda (LOGICAL_ERROR)

Detailed description / Documentation draft:
Fixes the following LOGICAL_ERROR:

    $ clickhouse-client -q 'select arrayFilter((a) -> ((a, arrayJoin([[]])) IN (Null, [Null])), [])'
    2020.08.16 00:32:01.967102 [ 1744189 ] {b40a5ebd-d710-4f03-bb18-57db67de1181} <Error> : Logical error: 'Lambda captured argument arrayJoin(array(array())) not found in required columns.'.
    clickhouse-server: ../src/Common/Exception.cpp:45: DB::Exception::Exception(const string&, int): Assertion `false' failed.

Since there are multiple input columns for arrayJoin():

    (gdb) p captured_names_
    $6 = std::vector of length 3, capacity 4 = {"arrayJoin(array(array()))", "arrayJoin(array(array()))", "__set"}

While FunctionCaptureOverloadResolver cannot handle non-unique columns.

Found with [AST fuzzer](https://clickhouse-test-reports.s3.yandex.net/10373/6c4748a63e7acde2cc3283d96ffec590aae1e724/fuzzer/fuzzer.log#fail1)

<details>

HEAD:
- 9cf246d208273500c493aed43f173db97ec02daf (before conflicts)

</details>

4ertus2 commented on src/Interpreters/ActionsVisitor.cpp:
```diff
@@ -430,7 +430,7 @@ size_t ScopeStack::getColumnLevel(const std::string & name)
 }
 
-void ScopeStack::addAction(const ExpressionAction & action)
+void ScopeStack::addAction(const ExpressionAction & action, bool addInput)
```
It looks like we could make a special `addActionNoInput()` instead of this flag. It's better to make a copy-paste than add any complexity here. No one understands this code even without a flag )

P.S. coding style: `add_input` instead of `addInput`.

azat replied:
>It looks like we could make a special addActionNoInput() instead of this flag. It's better to make a copy-paste than add any complexity here.

Ok

> No one understands this code even without a flag )

I thought that I was the only one )

azat commented:
> Fast test — fail: 1, passed: 1951, skipped: 234. Other tests won't be run before fast test will 

Will rebase after #13867

azat committed:
Fix arrayJoin() capturing in lambda

Fixes the following LOGICAL_ERROR:

    $ clickhouse-client -q 'select arrayFilter((a) -> ((a, arrayJoin([[]])) IN (Null, [Null])), [])'
    2020.08.16 00:32:01.967102 [ 1744189 ] {b40a5ebd-d710-4f03-bb18-57db67de1181} <Error> : Logical error: 'Lambda captured argument arrayJoin(array(array())) not found in required columns.'.
    clickhouse-server: ../src/Common/Exception.cpp:45: DB::Exception::Exception(const string&, int): Assertion `false' failed.

Since there are multiple input columns for arrayJoin():

    (gdb) p captured_names_
    $6 = std::vector of length 3, capacity 4 = {"arrayJoin(array(array()))", "arrayJoin(array(array()))", "__set"}

While FunctionCaptureOverloadResolver cannot handle non-unique columns.

alexey-milovidov commented:
01396_inactive_replica_cleanup_nodes - fixed in master.

alexey-milovidov commented:
AST Fuzzer: fixed in #12550
